### DISCUSSION TITLE
How to quickly find indices of edges that contain certain nodes from edge_index?

### DISCUSSION THREAD

**Discussion by scottshufe** [pyg-team/pytorch_geometric#5593] [Q&A] [answered]
Ask for help. My problem is just as the title described.
For example: 
Assume `edge_index=[[1, 2, 2, 3, 3], [2, 4, 5, 2, 5]]`,  and I want to find the indices of edges that contain `node_list=[2, 3]`. 
What's the quick solution for this question (without any for loops) ?

**Comment by EdisonLeeeee** [comment, marked as answer]
You can take a look at [torch_geometric.utils.subgraph](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.subgraph).

**Reply by scottshufe** [reply]
It seems that subgraph function cannot solve this problem.
Subgraph capture the edges in the graph consists of certain nodes, but cannot capture the edges that are not in this subgraph but their vertexes are in the subgraph. 
For example, 
```
edge_index = torch.tensor([[0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6],
                           [1, 0, 2, 1, 3, 2, 4, 3, 5, 4, 6, 5]])
subset = torch.tensor([3, 4, 5])
subgraph(subset, edge_index)
(tensor([[3, 4, 4, 5],
         [4, 3, 5, 4]]),)
```
I wanna capture all the edges contains nodes [3, 4, 5], however subgraph only returns the edges in the subgraph. Meanwhile, it cannot return the indices of these edges I seek for.
Anyway, really thanks for your suggestions.

**Reply by scottshufe** [reply]
Maybe I didn't express my problem clearly enough. The edges wanted are that one of vertexes is in a certain node list like [3, 4, 5].

**Reply by EdisonLeeeee** [reply]
Sorry I actually misunderstood that. But you can still achieve this with some necessary modifications on [torch_geometric.utils.subgraph](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.subgraph). Is that what you want?
```python
import torch
from torch_geometric.utils import subgraph, index_to_mask
from torch_geometric.utils.num_nodes import maybe_num_nodes

edge_index = torch.tensor([[0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6],
                           [1, 0, 2, 1, 3, 2, 4, 3, 5, 4, 6, 5]])
subset = torch.tensor([3, 4, 5])

num_nodes = maybe_num_nodes(edge_index)
subset = index_to_mask(subset, size=num_nodes)
edge_mask = subset[edge_index[0]] | subset[edge_index[1]]

>>> edge_index[:, edge_mask]
tensor([[2, 3, 3, 4, 4, 5, 5, 6],
        [3, 2, 4, 3, 5, 4, 6, 5]])
```
where `edge_mask` indicates the edges in the subgraph.

**Reply by scottshufe** [reply]
Thank you so much! It solves my problem perfectly.

**Reply by rusty1s** [reply]
Please also see the discussion here: https://github.com/pyg-team/pytorch_geometric/discussions/5595

**Reply by scottshufe** [reply]
Thanks. But here [#5593](https://github.com/pyg-team/pytorch_geometric/discussions/5593) is here 😂

**Reply by rusty1s** [reply]
Sorry, I meant https://github.com/pyg-team/pytorch_geometric/issues/5595.

**Reply by scottshufe** [reply]
Thank you so much.
